src/core_deku/ticket_table.ml

```diff
@@ -0,0 +1,106 @@
+open Helpers
+
+module Tickets = struct
+  module Map = Map.Make_with_yojson (Ticket_id)
+
+  type t = Amount.t Map.t [@@deriving yojson]
+
+  let get_balance t ~ticket = Map.find_opt ticket t
+
+  let of_seq = Map.of_seq
+
+  let singleton ~ticket ~amount = Map.of_seq (Seq.return (ticket, amount))
+
+  let merge t ~ticket ~amount =
+    let merger option ~amount =
+      let value =
+        Option.fold ~none:amount ~some:(fun x -> Amount.(x + amount)) option
+      in
+      Some value in
+    Map.update ticket (merger ~amount) t
+
+  let get_and_remove t ~ticket ~to_take =
+    let%ok amount =
+      Map.find_opt ticket t |> Option.to_result ~none:`Insufficient_funds in
+    let%assert () =
+      (`Insufficient_funds, Amount.to_int amount >= Amount.to_int to_take) in
+    let map =
+      Map.update ticket
+        (function
+          | Some amount ->
+            if Amount.equal amount to_take then
+              None
+            else
+              Some Amount.(amount - to_take)
+          | None -> assert false)
+        t in
+    Ok (to_take, map)
+
+  let subtract t ~ticket ~to_withdraw =
+    let%ok _, map = get_and_remove t ~ticket ~to_take:to_withdraw in
+    Ok map
+
+  let fetch t tickets =
+    let fetched =
+      List.fold_left_ok
+        (fun (lst, m) (ticket, amount) ->
+          let%ok amount_available, m =
+            get_and_remove m ~ticket ~to_take:amount in
+          Ok ((ticket, amount_available) :: lst, m))
+        ([], t) tickets in
+    fetched
+
+  let fold = Map.fold
+
+  let empty = Map.empty
+end
+
+module Map = Map.Make_with_yojson (Address)
+
+type t = Tickets.t Map.t [@@deriving yojson]
```

Comment:
I think `Tickets` could have a better name, a more descriptive one. But I don't have nothing in mind rn, so not gonna block it.

Reply:
renamed